### PR TITLE
Lag adjusts bioeffects

### PR DESCRIPTION
**Pull request by vode-code**
[FIX] [BALANCE]
## About the PR 
This PR lag adjusts all the OnLife() that I could find, with the exception of the Sims OnLife() , which appear to use an entirely different system. I would pay closer attention to the loops at Beneficial.dm line 235 and Harmful.dm line 715, the loops are directly rolling the odds multiple times, and there's probably some math that could make this cheaper.

## Why's this needed? 
Lag adjustment good.

## Changelog

```changelog
(u)Chickenish
(*)Mutations are now lag adjusted.
```

### CODE REVIEW

**Comment by Camryn-Buttes** [comment]
What about traits?

**Comment by ZeWaka** [comment]
labels pls

File: code/datums/abilities/revenant.dm [hunk 1]
```diff
@@ -172,17 +172,17 @@
...
 
-	OnLife()
+	OnLife(var/mult)
 		if (!src.wraith)
 			return
 		if (ghoulTouchActive)
-			ghoulTouchActive--
+			ghoulTouchActive -= mult
 			if (!ghoulTouchActive)
 				owner.show_message("<span class='alert'>You are no longer empowered by the netherworld.</span>")
```
Comment (Azrun) [inline code/datums/abilities/revenant.dm:181]: Need to bound ghoulTouchActive to [0-inf), or update `melee_attack_proc.dm` and here to check >0

File: code/modules/medical/genetics/bioEffects/beneficial.dm [hunk 2]
```diff
@@ -227,14 +227,17 @@
...
 		if(..()) return
 		var/mob/living/L = owner
-		L.HealDamage("All", heal_per_tick, heal_per_tick)
-		if (rand(1,regrow_prob) == 1 && ishuman(L))
-			var/mob/living/carbon/human/H = L
-			if (H.limbs)
-				H.limbs.mend(1)
+		L.HealDamage("All", heal_per_tick * mult, heal_per_tick * mult)
+		var/roundedmult
+		for (roundedmult = round(mult), roundedmult > 0, roundedmult --)
```
Comment (Azrun) [inline code/modules/medical/genetics/bioEffects/beneficial.dm:235]: Consider keep track of remainder of rounding, that way eventually the fractional delay will add up.  This does help resolve long pauses in OnLife though.

File: code/modules/medical/genetics/bioEffects/harmful.dm [hunk 3]
```diff
@@ -708,30 +708,29 @@
...
-		pulse.name = "emp pulse"
-		pulse.anchored = 1
-		SPAWN_DBG(2 SECONDS)
-			if (pulse) qdel(pulse)
-
-		//maybe have this only emp some things on the tile.
-		if(istype(T))
-			for (var/atom/O in T.contents)
-				O.emp_act()
+		for (roundedmult = round(mult), roundedmult > 0, roundedmult --)
```
Comment (Azrun) [inline code/modules/medical/genetics/bioEffects/harmful.dm:715]: Consider keeping track of remainder. 🤷

**Comment by vode-code** [comment]
> 
> 
> What about traits?

Didn't notice them, I had been searching for onLife(), not onLife(var/mob/owner), trying to avoid dealing with all the onLife(var/mult) that there already were.

**Comment by vode-code** [comment]
I think I got everything mentioned?

File: code/modules/medical/genetics/bioEffects/beneficial.dm [hunk 4]
```diff
@@ -227,14 +227,22 @@
...
 		if(..()) return
 		var/mob/living/L = owner
-		L.HealDamage("All", heal_per_tick, heal_per_tick)
-		if (rand(1,regrow_prob) == 1 && ishuman(L))
-			var/mob/living/carbon/human/H = L
-			if (H.limbs)
-				H.limbs.mend(1)
+		L.HealDamage("All", heal_per_tick * mult, heal_per_tick * mult)
+		var/roundedmult = round(mult)
+		var/roundedmultremainder
```
Comment (Azrun) [inline code/modules/medical/genetics/bioEffects/beneficial.dm:235]: Move to datum itself, so it is maintained for subsequent calls.

File: code/modules/medical/genetics/bioEffects/harmful.dm [hunk 5]
```diff
@@ -707,31 +707,35 @@
...
-			for (var/atom/O in T.contents)
-				O.emp_act()
+		for (roundedmult = roundedmult, roundedmult > 0, roundedmult --)
+			if (prob(50))
+				if (prob(5)) //don't really need this but to make it more harmful to the user.
+					T = get_turf(owner)
+				else
+					T = locate(owner.x + rand(-radius/2,radius+2), owner.y+rand(-radius/2,radius/2), 1)
+
+				var/obj/overlay/pulse = new/obj/overlay(T)
```
Comment (Azrun) [inline code/modules/medical/genetics/bioEffects/harmful.dm:727]: Consider checking if overlay already exists, stacking the effect multiple times doesn't seem necessarily.

File: code/modules/medical/genetics/bioEffects/harmful.dm [hunk 6]
```diff
@@ -791,12 +795,12 @@
...
 	icon_state  = "bad"
 
-	OnLife()
+	OnLife(var/mult)
 		if (ishuman(owner))
 			var/mob/living/carbon/human/H = owner
 
 			if (H.blood_volume > 400 && H.blood_volume > 0)
-				H.blood_volume -= 2
+				H.blood_volume -= 2*mult
```
Comment (Azrun) [inline code/modules/medical/genetics/bioEffects/harmful.dm:803]: This and polycythemia seem like an order of operations issue.  Unsure if this is a real issue though.

File: code/modules/medical/genetics/bioEffects/harmful.dm (hunk 5 above)
Comment (ZeWaka) [reply code/modules/medical/genetics/bioEffects/harmful.dm:727]: ^

File: code/modules/medical/genetics/bioEffects/harmful.dm (hunk 6 above)
Comment (vode-code) [reply code/modules/medical/genetics/bioEffects/harmful.dm:808]: this just extends the possible difference from 2 away from the thresholds to 2*mult away from the thresholds from what I can tell.

Comment (vode-code) [reply code/modules/medical/genetics/bioEffects/harmful.dm:808]: this is using lifeprocess mult, not life mult, so it'll be capped at 20u anyways.

File: code/modules/medical/genetics/bioEffects/harmful.dm [hunk 7]
```diff
@@ -707,31 +707,40 @@
...
-		//maybe have this only emp some things on the tile.
-		if(istype(T))
-			for (var/atom/O in T.contents)
-				O.emp_act()
+		for (roundedmult = roundedmult, roundedmult > 0, roundedmult --)
+			if (prob(50))
+				if (prob(5)) //don't really need this but to make it more harmful to the user.
+					T = get_turf(owner)
+				else
+					T = locate(owner.x + rand(-radius/2,radius+2), owner.y+rand(-radius/2,radius/2), owner.z)
```
Comment (Azrun) [inline code/modules/medical/genetics/bioEffects/harmful.dm:725]: Since this is location dependent it does seem a little goofy that you could then get a bunch of emps that burst when server gets behind for a while. 🤷

File: code/modules/medical/genetics/bioEffects/harmful.dm [hunk 8]
```diff
@@ -255,12 +255,12 @@
...
 
-	OnLife()
+	OnLife(var/mult)
 		if(..()) return
 		if (iscarbon(owner))
 			var/mob/living/carbon/C = owner
-			if (prob(tox_prob))
-				C.take_toxin_damage(tox_amount)
+			if (probmult(tox_prob))
+				C.take_toxin_damage(tox_amount*mult)
```
Comment (Tarmunora) [inline code/modules/medical/genetics/bioEffects/harmful.dm:263]: double-compensating

File: code/modules/medical/genetics/bioEffects/harmful.dm [hunk 9]
```diff
@@ -707,31 +707,40 @@
 	stability_loss = -10
 	var/const/radius = 2
 	icon_state  = "bad"
+	var/roundedmultremainder
 
-	OnLife()
+	OnLife(var/mult)
 		..()
-		if (prob(50))
-			return
```
Comment (Tarmunora) [inline code/modules/medical/genetics/bioEffects/harmful.dm:714]: Lag-compensation for this one could/should have just been adjusting this prob roll I feel...

File: code/modules/medical/genetics/bioHolder.dm [hunk 10]
```diff
@@ -544,14 +544,14 @@ var/list/datum/bioEffect/mutini_effects = list()
...
+	proc/OnLife(var/mult)
 		var/datum/bioEffect/BE
 		for(var/curr in effects)
 			BE = effects[curr]
 			if (BE)
-				BE.OnLife()
+				BE.OnLife(mult)
 				if(BE.timeLeft != -1)
-					BE.timeLeft--
+					BE.timeLeft -= mult
```
Comment (Tarmunora) [inline code/modules/medical/genetics/bioHolder.dm:554]:
```suggestion
					BE.timeLeft -= 1*mult
```

**Review by Tarmunora — CHANGES_REQUESTED** [review]

File: code/modules/medical/genetics/bioEffects/harmful.dm (hunk 9 above)
Comment (vode-code) [reply code/modules/medical/genetics/bioEffects/harmful.dm:714]: it's a lot more destructive with the ability to make multiple in one cycle from my testing.

Comment (Tarmunora) [reply code/modules/medical/genetics/bioEffects/harmful.dm:714]: Please change it to just adjust the prob roll

**Review by ZeWaka — CHANGES_REQUESTED** [review]
x

**Review by Tarmunora — APPROVED** [review]
